### PR TITLE
Add file extension filter

### LINKED ISSUE #2140: [FEATURE] introduce a file ending filter
Users might want to filter incoming updates based on the file endings on top of/instead of the mime type. Both attributes of documents are not safe, but that didn't stop us from introducing a mime type want and shouldn't stop us from a file ending one either. Both are a great way of initially filtering incoming files. We should just mention in the docs, as we do with mime types, that telegram doesnt offer any validation.

This filter can either be a subset of Filters.document or its own imo.

### PR DESCRIPTION
**Pull request by eIGato**
Resolve #2140.
Based on an abandoned PR of another contributor.
Fix #2156.

### CODE REVIEW

**Comment by eIGato** [comment]
@Bibo-Joshi may you restart failed tests? I don't think they failed because of these changes. More likely, some temporary issue with test env.
Besides, no need to edit keywords in this PR. "Resolve" and "Fix" are fine too.

**Comment by eIGato** [comment]
Thanks.

**Comment by Bibo-Joshi** [comment]
NP. I'll try to review both PRs tomorrow ;)

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:619]: What is this good for?

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:643]: why the hole splitting? `return message.document.file_name.endswith(f'.{self.file_extension}')` should do the trick

File: tests/test_filters.py
Comment (Bibo-Joshi) [inline tests/test_filters.py:428]: please add a test that makes sure the dot before the extension is respected, i.e. `file.foojpg` should not pass `Filters.file_extension('.jpg')`

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:628]: I do see why you want the user to not have to care about the dot, but I'd prefer to not fiddle with it and instead just document if the dot must be passed or not. Image someone for some reason wanting to filter for a `......ext` extension. That wouldn't work in the current setting.

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]

File: tests/test_filters.py
Comment (eIGato) [reply tests/test_filters.py:428]: I've already done it.

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:628]: This is the case i didn't think of. Okay.

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:619]: To not assing the attr in `__init__()` if it's None. Not much good, i'll might edit this.

Comment (Bibo-Joshi) [reply telegram/ext/filters.py:619]: please do ;)

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py]: It mattered when i `lstrip()`ed dots.

**Comment by eIGato** [comment]
I think everything is fixed here.

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:616]:
```suggestion
                ``Filters.document.file_extension('jpg')`` filters files with extension ``'.jpg'``.
```

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:631]: why is the `!r` necessary?

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:630]: my point was that this should only be either

* `self.file_extension = file_extension` and the docs should nota that the user needs to care abot the dot
* `self.file_extension = f'.{file_extension}'` and the docs should note that a dot will be added, not matter what

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
As a side note: Please don't force push if not necessary. It makes reviewing harder.

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:631]: Because it should differ for `None` and `"None"`.

**Comment by eIGato** [comment]
What about case sensitivity? I've made it insensitive but may be there are cases when it should not be done. For example `*.Dockerfile`. May be make it optional or just score on this?

**Comment by eIGato** [comment]
Sorry for force-pushing. I'm used to Gitlab where one can view diffs of different pushed versions. I just want to bring logically connected changes to one commit.

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [reply telegram/ext/filters.py]: but `str(None) == repr(None)`? o.0 shouldn't it rather be `f'{self.f_e if self.f_e else ""}'`, then?

**Comment by Bibo-Joshi** [comment]
> What about case sensitivity? I've made it insensitive but may be there are cases when it should not be done. For example `*.Dockerfile`. May be make it optional or just score on this?

Mh, good point. While intuitively I'd say lower case everything, a quick search tells me that file extension case sensitivity can depend on for operating system and also hard drive encoding or something. So I'd like to go for a flag, defaulting to lowercasing everythng. have a look at the implementation of `Filters.command` for that. Please add something like 

```
Note:
  * adds the dot automatically, so *insert .jpg example here*
  * case insensitive by default, use the flag *whatever it's called*, if you need it
```

to the docstring

> Sorry for force-pushing. I'm used to Gitlab where one can view diffs of different pushed versions. I just want to bring logically connected changes to one commit.

NP. We squash anyway before merging, so that doesn't really matter too much ;)

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py]: `str("None") != repr("None")` — that's what i mean. Please read the test to understand my intentions. May be i just got the task wrong. In short, `None` and `""` behaviors differ: `""` means filenames ending with a dot and `None` means empty filter that always returns `True` (except if there's no filename at all).

Comment (Bibo-Joshi) [reply telegram/ext/filters.py]: I do understand that `'None'` and `None` are not the same as input. Still I find `file_extension(None)` resulting in `'Filters.doucment.file_extension(None)'` a bis misleading as name (the difference to `"file_extension('None')"` being marginal and would prefer `'Filters.doucment.file_extension()'`. Anyway, I'm not too passionate about that, but it would maybe be good to explicitly ad tests checking that the name is set correctly

Comment (eIGato) [reply telegram/ext/filters.py]: Should we accept `None` at all? Argument is not optional, so `None` can only be passed explicitly. Adding `file_extension()` to the name may be misleading.

Comment (eIGato) [reply telegram/ext/filters.py]: Made it optional and removed `None` from name. But behavior question is still open. I've separated all `None`-related checks in one test, please read.

Comment (Bibo-Joshi) [reply telegram/ext/filters.py]: No, I think `None` should have to be passed explicitly. Though thinking about it, it would actually make sense to not just return `True` in that case but on the contrary, check that the file has *no* file extension, i.e. `return '.' not in file_extension`. Ofc that should be documented. 
In that case I do agree that `Filters.f_e(None)` would be a more appropriate naming.
I'll try to review this and the types PR later today

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:619]:
```suggestion
                    i.e. without a dot in the filename.
```

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [inline telegram/ext/filters.py:630]: I don't think the `*` is necessary

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the updates! Just two nitpicks left from my side. @Poolitzer do you want to review, too?

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:630]: Calling `Filter.file_extension('jpg', True)` is just insensible.

Comment (Bibo-Joshi) [reply telegram/ext/filters.py:630]: We don't use this kind of signatures anywhere else in the lib and I don't want to add it just in one place. As with `__init__` and lists: either change it everywhere (which is breaking and needs to be discussed), or not at all.
Aside from that, PTB is not about handholding its users.

**Comment by eIGato** [comment]
Cite from your contibuting guidelines: "This gives us the flexibility to re-order arguments and more importantly to add new required arguments".
Why shoul it refer only to calls but not the signatures?

**Comment by Bibo-Joshi** [comment]
> Cite from your contibuting guidelines: "This gives us the flexibility to re-order arguments and more importantly to add new required arguments".
> Why should it refer only to calls but not the signatures?

Well, that style guideline applies only to code within PTB itself and not to how users should use the library. So not having `, *,` in the signatures allows the user to use both.
I do see the advantage of enforcing kwargs in calls, but I also see the philosophy of not handholding the user and the fact that it's a breaking change that should be done consistently throughout the lib.
If it's important to you, please go ahead and open a new issue so that it can be properly discussed, as it's beyond the scope fo this PR.

**Comment by eIGato** [comment]
It's a normal approach: use a good practice for the new code and fix the old code later.
Did as you wish. Any nitpicks left?

**Comment by Bibo-Joshi** [comment]
not from my side, but Poolitzers review is still pending. Same for #2167 btw.

File: telegram/ext/filters.py
Comment (Poolitzer) [inline telegram/ext/filters.py:615]:
```suggestion
                    you may change this with the flag ``case_sensitive=True``.
```

this because it is a behaviour

File: telegram/ext/filters.py
Comment (Poolitzer) [inline telegram/ext/filters.py:619]: I would showcase that in an example

File: telegram/ext/filters.py
Comment (Poolitzer) [inline telegram/ext/filters.py:625]: this is a great example! Good job.

File: telegram/ext/filters.py
Comment (Poolitzer) [inline telegram/ext/filters.py:647]: sorry what is the !r for?

**Review by Poolitzer — CHANGES_REQUESTED** [review]
Left some minor comments, overall I really like this PR, thanks!

File: telegram/ext/filters.py
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:647]: it adds quotes to the strings. Also make `None` distinguishable from `"None"`. Also asked that :D

**Comment by Poolitzer** [comment]
oh @Bibo-Joshi can you check if you can build docs on this branch, specifically the filters page and this filter appears?

File: telegram/ext/filters.py
Comment (Poolitzer) [reply telegram/ext/filters.py:647]: quick google didnt show that to me :(

Comment (Bibo-Joshi) [reply telegram/ext/filters.py:647]: uses `repr()` instead of `str()`

**Comment by Bibo-Joshi** [comment]
> oh @Bibo-Joshi can you check if you can build docs on this branch, specifically the filters page and this filter appears?

It doesn't. @eIGato the docstring of `file_extension` needs to be copied into the docstring of `document`. Please do that and build the docs locally to check that it renders correctly.
thanks for the catch @Poolitzer !

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:619]: Done.

Comment (eIGato) [reply telegram/ext/filters.py:619]: Thanks.

File: telegram/ext/filters.py
Comment (eIGato) [reply telegram/ext/filters.py:625]: Thanks.

**Comment by eIGato** [comment]
IDKY the checks fail. New code is fully covered with tests and all tests pass locally: `1716 passed, 187 skipped, 19 xfailed, 1 xpassed in 207.14 seconds`.
I've tried to repush the branch to restart the checks but result did not change.

**Comment by Bibo-Joshi** [comment]
test official fails due to the recent API updates not yet being incorporated and codecov for some reason decided to mark a docstring as not tested, so everthing's fine.

**Comment by eIGato** [comment]
image
IDK how to fix such rendering: the renderer just understands the line feeds literally.

**Comment by Bibo-Joshi** [comment]
I fixed the rendering locally, but I can't push to your branch. did you disable edits by maintainers (very bottom of the panel on the right)? if not:

```python

        file_extension: This filter filters documents by their file ending/extension.

            Note:
                * This Filter only filters by the file ending/extension of the document,
                  it doesn't check the validity of document.
                * The user can manipulate the file extension of a document and
                  send media with wrong types that don't fit to this handler.
                * Case insensitive by default,
                  you may change this with the flag ``case_sensitive=True``.
                * Extension should be passed without leading dot
                  unless it's a part of the extension.
                * Pass :obj:`None` to filter files with no extension,
                  i.e. without a dot in the filename.

            Example:
                * ``Filters.document.file_extension("jpg")`` filters files with extension
                  ``".jpg"``.
                * ``Filters.document.file_extension(".jpg")`` filters files with extension
                  ``"..jpg"``.
                * ``Filters.document.file_extension("Dockerfile", case_sensitive=True)``
                  filters files with extension ``".Dockerfile"`` minding the case.
                * ``Filters.document.file_extension(None)`` filters files without a dot in the
                  filename.
```

**Review by Poolitzer — COMMENTED** [review]
LGTM with the rendering fixes bibo suggested

**Pull request merged**
